pkg/vault/contracts/AddUnbalancedLiquidityViaSwapRouter.sol

```diff
@@ -0,0 +1,228 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IPermit2 } from "permit2/src/interfaces/IPermit2.sol";
+
+import { IWETH } from "@balancer-labs/v3-interfaces/contracts/solidity-utils/misc/IWETH.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import {
+    IAddUnbalancedLiquidityViaSwapRouter
+} from "@balancer-labs/v3-interfaces/contracts/vault/IAddUnbalancedLiquidityViaSwapRouter.sol";
+import "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+import "@balancer-labs/v3-interfaces/contracts/vault/RouterTypes.sol";
+
+import { RouterHooks } from "./RouterHooks.sol";
+
+/**
+ * @notice Enable adding and removing liquidity unbalanced on pools that do not support it natively.
+ * @dev This extends the standard `Router` in order to call shared internal hook implementation functions.
+ * It factors out the unbalanced adds into two operations: a proportional add and a swap, executes them using
+ * the standard router, then checks the limits.
+ */
+contract AddUnbalancedLiquidityViaSwapRouter is RouterHooks, IAddUnbalancedLiquidityViaSwapRouter {
+    // Generalized SwapParams to accommodate both ExactIn and ExactOut.
+    struct SwapParams {
+        IERC20 tokenIn;
+        IERC20 tokenOut;
+        SwapKind kind;
+        uint256 amountGiven;
+        uint256 limit;
+        bytes userData;
+    }
+
+    constructor(
+        IVault vault,
+        IPermit2 permit2,
+        IWETH weth,
+        string memory routerVersion
+    ) RouterHooks(vault, weth, permit2, routerVersion) {
+        // solhint-disable-previous-line no-empty-blocks
+    }
+
+    /// @inheritdoc IAddUnbalancedLiquidityViaSwapRouter
+    function addUnbalancedLiquidityViaSwapExactIn(
+        address pool,
+        uint256 deadline,
+        bool wethIsEth,
+        AddLiquidityProportionalParams calldata addLiquidityParams,
+        SwapExactInParams calldata swapParams
+    ) external payable saveSenderAndManageEth returns (uint256[] memory amountsIn, uint256 swapAmountOut) {
```

Comment:
Here and elsewhere (applies to exact out and queries):
I think it'd be better to just return the actual amounts in, where:
- `amountsIn'[swapTokenIn] = amountsIn[swapTokenIn] + swapAmountIn`
- `amountsIn'[swapTokenOut] = amountsIn[swapTokenOut] - swapAmountOut`

Not sure whether it's really needed to split the two quantities for the caller. WDYT?